internal/internal_task_handlers.go

```diff
@@ -939,39 +939,31 @@ ProcessEvents:
 	// the replay of that event will panic on the decision state machine and the workflow will be marked as completed
 	// with the panic error.
 	var nonDeterministicErr error
+	var nonDeterminismType nonDeterminismDetectionType
 	if !skipReplayCheck && !w.isWorkflowCompleted || isReplayTest {
 		// check if decisions from reply matches to the history events
 		if err := matchReplayWithHistory(w.workflowInfo, replayDecisions, respondEvents); err != nil {
 			nonDeterministicErr = err
+			nonDeterminismType = nonDeterminismDetectionTypeReplayComparison
 		}
-	}
-	if nonDeterministicErr == nil && w.err != nil {
-		if panicErr, ok := w.err.(*workflowPanicError); ok && panicErr.value != nil {
-			if _, isStateMachinePanic := panicErr.value.(stateMachineIllegalStatePanic); isStateMachinePanic {
-				if isReplayTest {
-					// NOTE: we should do this regardless if it's in replay test or not
-					// but since previously we checked the wrong error type, it may break existing customers workflow
-					// the issue is that we change the error type and that we change the error message, the customers
-					// are checking the error string - we plan to wrap all errors to avoid this issue in client v2
-					nonDeterministicErr = panicErr
-				} else {
-					// Since we know there is an error, we do the replay check to give more context in the log
-					replayErr := matchReplayWithHistory(w.workflowInfo, replayDecisions, respondEvents)
-					w.wth.logger.Error("Ignored workflow panic error",
-						zap.String(tagWorkflowType, task.WorkflowType.GetName()),
-						zap.String(tagWorkflowID, task.WorkflowExecution.GetWorkflowId()),
-						zap.String(tagRunID, task.WorkflowExecution.GetRunId()),
-						zap.Error(nonDeterministicErr),
-						zap.NamedError("ReplayError", replayErr),
-					)
-				}
-			}
-		}
+	} else if panicErr, ok := w.getWorkflowPanicIfIllegaleStatePanic(); ok {
+		// This is a nondeterministic execution which ended up panicing
+		nonDeterministicErr = panicErr
+		nonDeterminismType = nonDeterminismDetectionTypeIllegalStatePanic
+		// Since we know there is an error, we do the replay check to give more context in the log
+		replayErr := matchReplayWithHistory(w.workflowInfo, replayDecisions, respondEvents)
+		w.wth.logger.Error("Illegal state caused panic",
+			zap.String(tagWorkflowType, task.WorkflowType.GetName()),
+			zap.String(tagWorkflowID, task.WorkflowExecution.GetWorkflowId()),
+			zap.String(tagRunID, task.WorkflowExecution.GetRunId()),
+			zap.Error(nonDeterministicErr),
+			zap.NamedError("ReplayError", replayErr),
+		)
 	}
 
 	if nonDeterministicErr != nil {
-
-		w.wth.metricsScope.GetTaggedScope(tagWorkflowType, task.WorkflowType.GetName()).Counter(metrics.NonDeterministicError).Inc(1)
+		scope := w.wth.metricsScope.GetTaggedScope(tagWorkflowType, task.WorkflowType.GetName(), tagNonDeterminismDetectionType, string(nonDeterminismType))
```

Comment:
```
string(nonDeterminismType)
```
tbh, the more I write Go, the more I want to get rid of all type casts (outside that type's methods) :|  there are just zero safety checks around them, feels like more than half of the time they just become landmines for later.  (int conversion, non-utf bytes, etc)

I'd probably just leave these as strings unless we have a typed API that accepts only them, because that's capable of preventing misuse.

Reply:
the underlying enum type alias is a string so this is actually safe but I agree on avoiding castings. however for metric tags enums are more clean way to define fixed set of values and requires this casting due to the tally api accepting strings only.

Reply:
I suppose this is actually in use due to the variable 🤔 
I do kinda think we'll want a more strongly typed API for metrics eventually (in particular: to break the dependency on a specific version of Tally), it'll be more useful then too.